src/EventStore.ClientAPI/Internal/ClusterDnsEndPointDiscoverer.cs

```diff
@@ -219,13 +215,13 @@ internal class ClusterDnsEndPointDiscoverer : IEndPointDiscoverer {
 							e = ae.Flatten();
 						_log.Error("Failed to get cluster info from [{0}]: request failed, error: {1}.", endPoint.EndPoint, e);
 					} finally {
-						completed.Set();
+						completedSource.TrySetResult(null);
 					}
 				},
 				// https://github.com/EventStore/EventStore/pull/2744#pullrequestreview-562358658
 				endPoint.V5HostHeader ? "" : endPoint.EndPoint.GetHost());
-			completed.Wait(_gossipTimeout);
-			return result;
+			Task.Delay(_gossipTimeout).ContinueWith(_ => completedSource.TrySetResult(null));
```

Comment:
i bumped on this, perhaps from the time i set a System.Threading.Timer and it got garbage collected and stopped calling its callback -- but i think what you've got is safe